drivers/adc/adc_smartbond_gpadc.c

```diff
@@ -251,26 +302,79 @@ static int adc_smartbond_read_async(const struct device *dev,
 }
 #endif /* CONFIG_ADC_ASYNC */
 
-static int adc_smartbond_init(const struct device *dev)
+static int gpadc_smartbond_resume(const struct device *dev)
 {
-	int err;
-	struct adc_smartbond_data *data = dev->data;
+	int ret;
 	const struct adc_smartbond_cfg *config = dev->config;
 
-	GPADC->GP_ADC_CTRL_REG = GPADC_GP_ADC_CTRL_REG_GP_ADC_EN_Msk;
-	GPADC->GP_ADC_CTRL2_REG = 0;
-	GPADC->GP_ADC_CTRL3_REG = 0x40;
-	GPADC->GP_ADC_CLEAR_INT_REG = 0x0;
+	da1469x_pd_acquire(MCU_PD_DOMAIN_PER);
 
 	/*
 	 * Configure dt provided device signals when available.
 	 * pinctrl is optional so ENOENT is not setup failure.
 	 */
-	err = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
-	if (err < 0 && err != -ENOENT) {
-		LOG_ERR("ADC pinctrl setup failed (%d)", err);
-		return err;
+	ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0 && ret != -ENOENT) {
+		LOG_ERR("ADC pinctrl setup failed (%d)", ret);
+		return ret;
```

Comment:
Should we release PD_PER in case of error?

Reply:
Yes, PD_PER should be released. What is more `gpadc_smartbond_pm_action` needs update so it returns the error code of `gpadc_smartbond_resume` and `gpadc_smartbond_suspend`. Now, the callback function always returns success and so in case of error PM or driver should not be aware of.